[quant] Support for fused ConvBn1d and ConvBnRelu1d modules

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38452 [quant] Support for fused ConvBn1d and ConvBnRelu1d modules**

Summary:

Test Plan:
python test/test_quantization.py TestFused

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21632878](https://our.internmc.facebook.com/intern/diff/D21632878)

supriyar committed:
[quant] Support for fused ConvBn1d and ConvBnRelu1d modules
Summary:

Test Plan:
python test/test_quantization.py TestFused

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support for fused ConvBn1d and ConvBnRelu1d modules"


Summary:

Test Plan:
python test/test_quantization.py TestFused

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support for fused ConvBn1d and ConvBnRelu1d modules"


Summary:

Test Plan:
python test/test_quantization.py TestFused

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/fuse_modules.py:
```diff
@@ -56,9 +57,14 @@ def fuse_conv_bn_relu(conv, bn, relu):
-        return torch_fused.ConvReLU3d(
-            torch.nn.utils.fusion.fuse_conv_bn_eval(conv, bn), relu) if is_3d \
-            else torch_fused.ConvReLU2d(
+        if is_1d:
```
How about 
```
map_to_fused_module = {
torch.nn.Conv1d: torch_fused.ConvReLU1d,
torch.nn.Conv2d: torch_fused.ConvReLU2d,
torch.nn.Conv3d: torch_fused.ConvReLU3d,
}

fused_module = map_to_fused_module[type(conv)]
return fused_module(torch.nn.utils.fusion.fuse_conv_bn_eval(conv, bn), relu)
```

and remove all the is_1d/2d checks?

raghuramank100 commented on torch/quantization/fuse_modules.py:
```diff
@@ -48,6 +48,7 @@ def fuse_conv_bn_relu(conv, bn, relu):
     assert(conv.training == bn.training == relu.training),\
         "Conv and BN both must be in the same mode (train or eval)."
     is_3d = isinstance(conv, torch.nn.Conv3d)
+    is_1d = isinstance(conv, torch.nn.Conv1d)
```
For the training part, you can do this:
map_to_fused_module_train = {
torch.nn.Conv2d: torch_fused.ConvReLU2d,
torch.nn.Conv3d: torch_fused.ConvReLU3d,
}
fused_module = map_to_fused_module_train.get(type(conv))
if fused_module is not None:
      return fused_module(conv,bn,relu)
else:
      raise NotImplementedError("Cannot fuse modules: {}".format((conv, bn, relu)))

raghuramank100 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -631,6 +631,10 @@ def __init__(self, qconfig):
         self.bn2 = nn.BatchNorm3d(2).to(dtype=torch.float)
         self.relu3 = nn.ReLU(inplace=True).to(dtype=torch.float)
         self.conv3 = nn.Conv1d(3, 3, 2).to(dtype=torch.float)
+        if train:
```
Can you instead check that the code raises an exception for Conv1d + Bn1d and remove the train flag for the model?

raghuramank100 commented on test/quantization/test_quantize.py:
```diff
@@ -1259,7 +1259,7 @@ def checkQuantized(model):
 class TestFusion(QuantizationTestCase):
     def test_fuse_module_train(self):
-        model = ModelForFusion(default_qat_qconfig).train()
+        model = ModelForFusion(default_qat_qconfig, train=True).train()
```
Please see comment below on removing train as an option for building the model

supriyar committed:
Update on "[quant] Support for fused ConvBn1d and ConvBnRelu1d modules"


Summary:

Test Plan:
python test/test_quantization.py TestFused

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support for fused ConvBn1d and ConvBnRelu1d modules"


Summary:

Test Plan:
python test/test_quantization.py TestFused

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support for fused ConvBn1d and ConvBnRelu1d modules"


Summary:

Test Plan:
python test/test_quantization.py TestFused

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21632878](https://our.internmc.facebook.com/intern/diff/D21632878)

[ghstack-poisoned]
